framework/src/bundle/CoreBundleContext.h

```diff
@@ -187,6 +198,31 @@ namespace cppmicroservices
 
         void Uninit1();
 
+        /**
+         * Called when framework shutdown has begun.
+         * This blocks (while returned object is held):
+         *     - other calls to start or stop the framework
+         *     - calls to start bundles
+         */
+        WriteLock BlockDuringFrameworkShutdown();
+
+        /**
+         * Called when framework startup has begun.
+         * This blocks (while returned object is held):
+         *     - other calls to start or stop the framework
+         *     - calls to start bundles
+         */
+        WriteLock BlockDuringFrameworkStartup();
```

Comment:
since these methods are setting state and returning a way to guard against other threads modifying the framework state, what if they are called:
```
SetFrameworkStateToStoppedAndBlockUntilComplete
SetFrameworkStateToStartedAndBlockUntilComplete
```
alternatively, you could pass in a bool to one method named:
```
SetFrameworkStateAndBlockUntilComplete
```

yes, these methods are wordy, but I think its worth it. Also, this solution surfaces a more widespread issue with how we start and stop the Framework and ensure that concurrent public Framework API calls function without crashing. I would eventually like to see a state machine type solution, similar to what is done in Declarative Services. however that would take more time and effort and this is a reasonable solution for now.